### PR TITLE
fix "container is already disposed" in tests

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Problem
`ProjectRule` will automatically dispose `Project` object every time finishing a test. However we have some time lagging API calls which are invoked during the tests in the background thread or scheduled for a later call and still keep references to the disposed `Project` objects. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/startup/CodeWhispererProjectStartupSettingsListener.kt [hunk 1]
```diff
@@ -54,6 +54,8 @@ class CodeWhispererProjectStartupSettingsListener(private val project: Project)
     }
 
     override fun activeConnectionChanged(newConnection: ToolkitConnection?) {
+        // TODO: investigate why the project level listener not being dispoased when its parent gets disposed
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/startup/CodeWhispererProjectStartupSettingsListener.kt:57]: can you log the full name of the class when this happens to make this easier to investigate?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt [hunk 2]
```diff
@@ -58,10 +62,15 @@ fun <T> calculateIfIamIdentityCenterConnection(connection: ToolkitConnection, ca
 // Controls the condition to send telemetry event to CodeWhisperer service, currently:
 // 1. It will be sent for Builder ID users, only if they have optin telemetry sharing.
 // 2. It will be sent for IdC users, regardless of telemetry optout status.
-fun runIfIdcConnectionOrTelemetryEnabled(project: Project, callback: (connection: ToolkitConnection) -> Unit) =
+fun runIfIdcConnectionOrTelemetryEnabled(project: Project, callback: (connection: ToolkitConnection) -> Unit) {
+    if (project.isDisposed) {
+        return
+    }
```
Comment (andrewyuq) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt:68]: same don't need here.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt [hunk 3]
```diff
@@ -44,8 +44,12 @@ import software.aws.toolkits.telemetry.CodewhispererCompletionType
...
 
 fun <T> calculateIfIamIdentityCenterConnection(project: Project, calculationTask: (connection: ToolkitConnection) -> T): T? =
-    ToolkitConnectionManager.getInstance(project).activeConnectionForFeature(CodeWhispererConnection.getInstance())?.let {
-        calculateIfIamIdentityCenterConnection(it, calculationTask)
+    if (project.isDisposed) {
+        null
+    } else {
+        ToolkitConnectionManager.getInstance(project).activeConnectionForFeature(CodeWhispererConnection.getInstance())?.let {
+            calculateIfIamIdentityCenterConnection(it, calculationTask)
+        }
```
Comment (andrewyuq) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt:52]: I don't think we will need this, the above check in `activeConnectionChanged` should be good. I prefer not to hide too many issues if there are at the same time.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt (hunk 2 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt:68]: sg

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt (hunk 3 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererUtil.kt:52]: sg

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/startup/CodeWhispererProjectStartupSettingsListener.kt (hunk 1 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/startup/CodeWhispererProjectStartupSettingsListener.kt:57]: sg

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt [hunk 4]
```diff
@@ -56,6 +56,7 @@ private fun notifyNewCustomization(project: Project) {
     )
 }
 
+// TODO: better to split this component into two parts, of which one is application level component and one is project level component in the long run
```
Comment (Will-ShaoHua) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: just like how ToolkitAuthManager and ToolkitConnectionManager work together

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt [hunk 5]
```diff
@@ -219,36 +220,40 @@ class DefaultCodeWhispererModelConfigurator : CodeWhispererModelConfigurator, Pe
...
-                if (forceUpdate) {
+                null -> run {
                     ApplicationManager.getApplication().executeOnPooledThread {
                         // will update devTool tree
-                        val updatedValue = listCustomizations(project, passive = true) != null
-                        if (updatedValue != cachedValue) {
-                            project.refreshDevToolTree()
-                        }
+                        listCustomizations(project, passive = true) != null
+                        project.refreshDevToolTree()
```
Comment (andrewyuq) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:235]: do we still need the `!= null`?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt (hunk 4 above)
Comment (andrewyuq) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: I still think CodeWhisperer should be an application service overall not project one.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt (hunk 5 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:235]: nope, should remove

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt (hunk 4 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: yes, so I think like how ToolkiAuth & ToolkitConnection managers work, the drawback of current implementation is that the customization selection is bound to "connection" level, it's not easy to make it "project" level binding with current implementation. But I will take this as follow-up and it's not urgently required for now.

Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: My idea is that 
App level component will manage the key-value pair of conneciton to customization options snapshot, validation etc. 


Project level component will simply manage the selection of customization.

Comment (andrewyuq) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: sure,
but my point is I would just make CW connection a app-level connection.

**Review by andrewyuq — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt (hunk 4 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: I see

Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/customization/CodeWhispererModelConfigurator.kt:59]: > but my point is I would just make CW connection a app-level connection.
that would make things easier

**Pull request merged**